Improve a few aspects of the install script

lbt opened the pull request:
There are a few commits which I thought would make the script more useful.

In particular I wanted to be able to hack on the code so installing 'editable' from a local clone rather than using pip install directly from git. This also allows the local repos to be re-checked-out to specific versions to solve pip dependency issues.

lbt committed:
Exit if install not wanted

Rather than having all the install stuff inside an if/then

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Set enable-linger for the user

This ensures that all the user services are launched at boot by
systemd.

See https://www.freedesktop.org/software/systemd/man/loginctl.html :

enable-linger [USER…], disable-linger [USER…]

    Enable/disable user lingering for one or more users. If enabled
    for a specific user, a user manager is spawned for the user at
    boot and kept around after logouts. This allows users who are not
    logged in to run long-running services. Takes one or more user
    names or numeric UIDs as argument. If no argument is specified,
    enables/disables lingering for the user of the session of the
    caller.

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Use sudo consistently

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Allow ovos source to be cloned and used locally

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Exit on error

If something goes wrong just stop. The script should be re-runnable
and it allows the user to see issues at once rather than having
scroll back and figure out what went wrong.

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Only start pulseaudio if it's not running

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Create and install into a venv

Using a venv rather than a 'user install' makes it much easier to
contain the python environment and even have multiple versions of OVOS
installed at once.

It also solves the problem of root being able to run certain services
since the venv can easily be invoked by root using the venv's python3
executable. (Currently this is a bit hardcoded however there are more
elegant solutions that can be implemented in time).

From https://docs.python.org/3/library/venv.html :
"You don’t specifically need to activate a virtual environment, as you
can just specify the full path to that environment’s Python
interpreter when invoking Python. Furthermore, all scripts installed
in the environment should be runnable without activating it."

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Save the answers to the questions to allow easy re-running

OK, to be fair this is mainly for debugging this script!

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Don't set PATH

There's no point. It doesn't persist once the script has run,
nothing below needs it and no daemons inherit it (and it'd be a bug if
they relied on it!)

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Install everything at once with pip where possible

Installing all at once allows pip to check and download all
dependencies in one go and to give nice sane messages about version
conflicts

This mechanism also allows some repo/package groups to be omitted.

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Padatious is needed and fann2 version must be fixed for raspbian

Signed-off-by: David Greaves <david@dgreaves.com>

Bulwagga commented:
These are nice ideas! One of my own goals for the script is that it will allow people to install OVOS onto a working Raspberry Pi image without needing to rebuild any existing environment.  I have a few dedicated systems that act as servers and data monitors and I'd like to give them voice control.  Anything in your changes that would cause problems for such systems?

lbt commented:
No. That's my goal too. Things like using the venv and a dedicated user will support good isolation.

My next step is to create a remote dinkum-listener which will run on a few machines around the house that are already doing other things :)

Bulwagga commented:
Got it.  Vision was also to keep things as simple as possible for development testing and getting new users up to speed quickly. Things like creating the python groups would make it easier for new users to understand the dependencies required when choosing among several listeners, for example. Builderjer can speak to the development environment considerations, I am coming from the user side.

lbt committed:
Only clone repos that we install if we're doing local hacking

For example setting many values to " " to disable them (annoyingly
bash doesn't distinguish between set to null and unset for ${VAR:=} )

export OVOS_CORE_REPOS=" " OVOS_TTS_REPOS=" " OVOS_EXTRA_REPOS=" "
OVOS_PHAL_REPOS=" " OVOS_SKILLS_REPOS=" " OVOS_EXTRA_SKILL_REPOS=" "
OVOS_DINKUM_REPOS= OVOS_PRECISE_LITE_REPOS= OVOS_STT_REPOS=

Now a run of namual-setup.sh will only clone the repos needed to run a
remote satellite listener.

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
It's good practice to upgrade pip and wheel in a venv too

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Document manual-setup.sh and modifying modules to run a satellite

Signed-off-by: David Greaves <david@dgreaves.com>

builderjer reviewed (commented):
I realize that the README could use some updating, but I don't think it should include the satellite option like that.  Maybe put it in the `manual_user_install.sh` script as an option?

builderjer commented on manual_user_install.sh:
```diff
@@ -1,73 +1,76 @@
-    pip3 install git+https://github.com/OpenVoiceOS/ovos-ww-plugin-precise-lite
-    pip3 install --upgrade setuptools
+    # Precise-lite wake-word (ww) cluster
+    pip3 install --upgrade setuptools wheel pip
```
Should this be moved uphill? As in the first thing installed and upgraded?

lbt replied:
yes - I just kept things in pretty much the same order initially

builderjer commented on manual_user_install.sh:
```diff
@@ -146,30 +153,20 @@ function install_extra_skills (){
-    pip3 install git+https://github.com/OpenVoiceOS/skill-ovos-news
-    pip3 install git+https://github.com/OpenVoiceOS/skill-ovos-somafm
-    pip3 install git+https://github.com/OpenVoiceOS/skill-ovos-youtube-music
+    # Add the NeonGeckoCom skills
```
I think I want to remove the Neon skills from the image.  Most are in OVOS repos, and the ones that aren't don't play well with OVOS all of the time.  And they bring in extra dependencies such as `neon-utils`

lbt replied:
OK - I didn't want to make any functional changes at this point

builderjer replied:
In fact, I did remove them from the image.  They are just still in the `install` script

Bulwagga replied:
I leave the Neon skills out of my builds... On a 3B+ I want to be as lightweight as possible. I didn't realize they were pulled from the image.

builderjer commented on manual_user_install.sh:
```diff
@@ -189,84 +186,141 @@ echo
+echo ""
+echo "1. Consider creating an .asoundrc and check your microphone with alsamixer, arecord, and aplay."
+echo ""
+echo "2. You can find documentation at https://openvoiceos.github.io/community-docs/install_raspbian/"
```
This link is dead.  Maybe just point to the main page of the docs?

lbt replied:
yes, I didn't write this - just changed the whitespace :)

builderjer commented on manual_user_install.sh:
```diff
@@ -189,84 +186,141 @@ echo
+echo ""
+echo "2. You can find documentation at https://openvoiceos.github.io/community-docs/install_raspbian/"
+echo ""
+echo "3. You can find pre-built OVOS/PI images at https://ovosimages.ziggyai.online/raspbian/"
```
This is a wrong link also
should be `https://ovosimages.ziggyai.online/raspbian/development`
The way you have it shows several other image options also, which I guess is alright

lbt replied:
same - whitespace change :)

Bulwagga commented:
This has fallen behind during some exploratory work on admin-phal.

builderjer commented:
I just changed the admis-phal service again. This is the only way I can get it to work correctly

Bulwagga commented:
Agree on the new admin-phal. This was the only substantial change in my most recent pull request (other than the links, etc.). I will cancel my PR to keep things simpler as you pull in jtb's work.

builderjer commented:
If this works for you, @Bulwagga , I will merge with the suggested changes to the Readme especially

Bulwagga commented:
Absolutely.  These seem like useful enhancements to me. In the meantime I'm giving thought to the ways raspovos is going to be used, and what features would be most useful for each.  Clearly there are developer users and we now have two of us that are interested in using OVOS an an overlay to support existing systems.  I can see a third cluster of users that simply want a smart device that protects their information.I'm thinking we want avoid building a "spork".  All the spearing capacity of the spoon  combined with the soup slurping capacity of the fork.From the outside looking in, I can see OVOS deeply embedded into the smart device if only to be lean and mean. At the same time I can see the overlay as a compartmentalized installation with something like venv. A dev environment would need maximum flexibility, not sure how to architect that.Then prioritize development tasks from those product descriptions.I'm excited to see these changes going in!+}
-------- Original message --------From: builderjer ***@***.***> Date: 8/24/23  7:25 PM  (GMT-05:00) To: OpenVoiceOS/raspbian-ovos ***@***.***> Cc: Bulwagga ***@***.***>, Mention ***@***.***> Subject: Re: [OpenVoiceOS/raspbian-ovos] WIP - improve a few aspects of the
  install script (PR #76) 
If this works for you, @Bulwagga , I will merge with the suggested changes to the Readme especially

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

builderjer commented:
I am going to merge this after the suggested changes, and this discussion needs to be addressed more [here](https://github.com/orgs/OpenVoiceOS/discussions/185)

builderjer commented on README.md:
```diff
@@ -14,6 +14,27 @@ This guide was originally designed for a headless, (No GUI) Raspberry Pi 3. The
+
+Run `./manual_user_install.sh` for a full install which will work locally
+
+To just install a partial set of the modules set the unneeded ones to a space in the environment. This example will install only the modules for a satellite listener:
```
This section should be removed from the README.  Connecting this way, while works, is strongly discouraged.

lbt committed:
Make installing setuptools/wheel/pip part of venv creation

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Remove Neon skills

As suggested by builderjet/Bulwaga

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Revert "Document manual-setup.sh and modifying modules to run a satellite"

This reverts commit 15024769cf28ae14bfc94a5220bf835c2c83fa70.

lbt committed:
Fix prebuilt image url

Signed-off-by: David Greaves <david@dgreaves.com>

lbt committed:
Changed documentation link

Signed-off-by: David Greaves <david@dgreaves.com>

lbt commented:
Let me know if I missed anything @builderjer

builderjer commented:
Looks good!!

Thanks for the help
